pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -618,31 +621,100 @@ func (c *Cluster) saveMonConfig() error {
 	d := c.makeDeployment(m, hostname)
-	logger.Debugf("Starting mon: %+v", d.Name)
-	_, err := c.context.Clientset.AppsV1().Deployments(c.Namespace).Create(d)
+	var deploymentExists bool
+	_, err := c.context.Clientset.AppsV1().Deployments(c.Namespace).Get(
+		d.Name, metav1.GetOptions{})
+	if err == nil {
+		deploymentExists = true
+	} else {
+		if errors.IsNotFound(err) {
+			deploymentExists = false
+		} else {
+			return fmt.Errorf("failed to get mon deployment %s. %+v", d.Name, err)
+		}
+	}
+
+	// if the deployment exists, also check if it is using PVC storage
+	pvc, err := c.makeDeploymentPVC(m)
 	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create mon deployment %s. %+v", m.ResourceName, err)
+		return fmt.Errorf("failed to make mon %s pvc. %+v", d.Name, err)
+	}
+	var pvcExists bool
+	if deploymentExists {
+		_, err = c.context.Clientset.CoreV1().PersistentVolumeClaims(
+			c.Namespace).Get(pvc.Name, metav1.GetOptions{})
+		if err == nil {
+			pvcExists = true
+		} else if errors.IsNotFound(err) {
+			pvcExists = false
+		} else {
+			return fmt.Errorf("failed to get pvc %s. %+v", pvc.Name, err)
 		}
-		logger.Infof("deployment for mon %s already exists. updating if needed", m.ResourceName)
+	} else {
+		pvcExists = false
+	}
+
+	// persistent storage is not altered after the deployment is created. this
+	// means we need to be careful when updating the deployment to avoid new
+	// changes to the crd to change an existing pod's persistent storage. the
+	// deployment spec created above does not specify persistent storage. here
+	// we add in PVC or HostPath storage based on an existing deployment OR on
+	// the current state of the CRD.
+	if pvcExists || (!deploymentExists && c.spec.Mon.VolumeClaimTemplate != nil) {
+		d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes,
+			opspec.DaemonVolumesDataPVC(pvc.Name))
+	} else {
+		d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes,
+			opspec.DaemonVolumesDataHostPath(m.DataPathMap)...)
+	}
+
+	if deploymentExists {
+		logger.Infof("deployment for mon %s already exists. updating if needed",
+			d.Name)
 
-		// Always invoke ceph version before an upgrade so we are sure to be up-to-date
 		daemonType := string(config.MonType)
 		var cephVersionToUse cephver.CephVersion
-		currentCephVersion, err := client.LeastUptodateDaemonVersion(c.context, c.clusterInfo.Name, daemonType)
+		currentCephVersion, err := client.LeastUptodateDaemonVersion(c.context,
+			c.clusterInfo.Name, daemonType)
 		if err != nil {
-			logger.Warningf("failed to retrieve current ceph %s version. %+v", daemonType, err)
-			logger.Debug("could not detect ceph version during update, this is likely an initial bootstrap, proceeding with c.clusterInfo.CephVersion")
+			logger.Warningf("failed to retrieve current ceph %s version. %+v",
+				daemonType, err)
+			logger.Debug("could not detect ceph version during update, this " +
```

Comment:
does your editor break up the strings? i tend to lean towards longer strings to keep them together, but not a big deal. it also makes the diff look bigger if nothing changed otherwise.

Reply:
I didn't intend to leave these (existing) reformatted strings this way, so I'll put them back to make the diff simpler.

I use a terminal editor sized to around 80 columns. This definitely influences new code I'm writing, but this is almost entirely a non-technical personal preference. Perhaps Rook can adopt some standards or guidelines?